File: R/readsnowsig.R
#
# NAME:
# readsnowsig
#
# PURPOSE:
# To read collocate HDF4 files containing validation data for signature
# studies.
#
# NOTES:
# Not fully finished... 
#
# BUGS:
# Probably in the underlying C-function...
# C functions does not transfer strings back properly...
#
# AUTHOR:
# Øystein Godøy, MET/FOU, 30.01.2003
#
# MODIFIED:
# Øystein Godøy, met.no/FOU, 20.11.2004
# Added extraction of NWP and slightly modified data reduction...
#
# CVS_ID:
# $Id: readsnowsig.R,v 1.3 2013-04-12 10:29:24 steingod Exp $
#

readsnowsig <- function(filename,classname="snow",station="NA",start="NA",end="NA") {

    if (missing(filename)) {
        cat("Husk at filnavn må oppgis...\n")
            return;
    }

    start <- as.POSIXct(strptime(start,format="%d%b%Y"))
        end <- as.POSIXct(strptime(end,format="%d%b%Y"))
        if (is.na(start)) start <- 0
            if (is.na(end)) end <- 0

                noobs <- 0

                    tmp <- .C("checkrec",
                            filename=as.character(filename),
                            noobs=as.integer(noobs),
                            classname=as.character(classname),
                            station=as.character(station),
                            start=as.integer(start),end=as.integer(end))
                    if (tmp$noobs <= 0) return

                        nopix <- 13*13
                            size <- tmp$noobs*nopix
                            tid <- vector(mode="integer",length=size)
                            stid <- vector(mode="integer",length=size)
                            N <- vector(mode="integer",length=size)
                            CL <- vector(mode="integer",length=size)
                            CM <- vector(mode="integer",length=size)
                            CH <- vector(mode="integer",length=size)
                            E <- vector(mode="integer",length=size)
                            sss <- vector(mode="integer",length=size)
                            k1 <- real(length=size)
                            k2 <- real(length=size)
                            k3a <- real(length=size)
                            k3b <- real(length=size)
                            k4 <- real(length=size)
                            k5 <- real(length=size)
                            soz <- real(length=size)
                            saz <- real(length=size)
                            raz <- real(length=size)
                            t0m <- real(length=size)
                            t2m <- real(length=size)

                            tmp <- .C("readsnowsig",
                                    filename=as.character(filename),
                                    noobs=as.integer(tmp$noobs),nopix=as.integer(nopix),
                                    station=as.character(station),
                                    start=as.integer(start),end=as.integer(end),
                                    stid=as.integer(stid),tid=as.integer(tid),
                                    N=as.integer(N),
                                    CL=as.integer(CL), CM=as.integer(CM), CH=as.integer(CH),
                                    E=as.integer(CL), sss=as.integer(CM),
                                    k1=as.real(k1), k2=as.real(k2), k3a=as.real(k3a),
                                    k3b=as.real(k3b),
                                    k4=as.real(k4), k5=as.real(k5), 
                                    soz=as.real(soz), saz=as.real(saz), raz=as.real(raz),
                                    t0m=as.real(t0m), t2m=as.real(t2m)
                                    )

# Perform data transformations
                            tmp$tid <- ISOdate(1970,1,1)+tmp$tid 
                            tmp$k1[tmp$k1 < -100] <- NA
                            tmp$k2[tmp$k2 < -100] <- NA
                            tmp$k3a[tmp$k3a < -100] <- NA
                            tmp$k3b[tmp$k3b < -100] <- NA
                            tmp$k4[tmp$k4 < -100] <- NA
                            tmp$k5[tmp$k5 < -100] <- NA
                            tmp$soz[tmp$soz < -100] <- NA
                            tmp$saz[tmp$saz < -100] <- NA
                            tmp$raz[tmp$raz < -100] <- NA
                            tmp$N[tmp$N < 0] <- NA
                            tmp$CL[tmp$CL < 0] <- NA
                            tmp$CM[tmp$CM < 0] <- NA
                            tmp$CH[tmp$CH < 0] <- NA
                            tmp$E[tmp$E < 0] <- NA
                            tmp$sss[tmp$sss < 0] <- NA

# Create data.frame for easier data handling during analysis
                            dafr <- data.frame(
                                    tid=tmp$tid[tmp$soz<85.],
                                    N=tmp$N[tmp$soz<85.],
                                    CL=tmp$CL[tmp$soz<85.],
                                    CM=tmp$CM[tmp$soz<85.],
                                    CH=tmp$CH[tmp$soz<85.],
                                    E=tmp$E[tmp$soz<85.],
                                    sss=tmp$sss[tmp$soz<85.],
                                    k1=tmp$k1[tmp$soz<85.],
                                    k2=tmp$k2[tmp$soz<85.],
                                    k3=tmp$k3[tmp$soz<85.],
                                    k4=tmp$k4[tmp$soz<85.],
                                    k5=tmp$k5[tmp$soz<85.],
                                    soz=tmp$soz[tmp$soz<85.],
                                    saz=tmp$saz[tmp$soz<85.],
                                    raz=tmp$raz[tmp$soz<85.],
                                    t0m=tmp$t0m[tmp$soz<85.],
                                    t2m=tmp$t2m[tmp$soz<85.])

                            return(dafr)
}

File: src/readsnowsig.c
/*
 * NAME:
 * readsnowsig.c
 *
 * PURPOSE:
 * To dump data stored in NCSA HDF5 files by software collocate...
 *
 * REQUIREMENTS:
 * NA
 *
 * INPUT:
 * NA
 *
 * OUTPUT:
 * NA
 *
 * NOTES:
 * NA
 *
 * BUGS:
 * NA
 *
 * AUTHOR:
 * Øystein Godøy, MET/FOU, 21.01.2003
 *
 * MODIFIED:
 * Øystein Godøy, met.no/FOU, 20.11.2004: Added collection of NWP data...
 * Øystein Godøy, METNO/FOU, 2013-04-12: Changed usage of header files.
 *
 * CVS_ID:
 * $Id: readsnowsig.c,v 1.2 2013-04-12 10:29:24 steingod Exp $
 */

#ifdef HAVE_LIBHDF5

#include <readfmcol.h>
#include <unistd.h>
#include <dirent.h>
#include <time.h>
#include <utime.h>
#include <sys/types.h>
#include <sys/stat.h>

/*
   int sm_debug=1;
   */

void readsnowsig(char **infile, int *n, int *p, char **station, 
        int *start, int *end,
        int *stid, int *mytime,
        int *N, int *CL, int *CM, int *CH, 
        int *E, int *sss, 
        double *k1, double *k2, 
        double *k3a, double *k3b, 
        double *k4, double *k5,
        double *soz, double *saz, double *raz,
        double *t0m, double *t2m
        ) {

    /* General variables */
    char *where="readsnowsig";
    int i, j, k;
    int status, noobs;
    obsstruct *sobs;
    as_data *asdata;
    ns_data *nsdata;
    skeys scrit = {"snow","NA",0,0};
    struct tm ut;
    time_t vt, mymax=0, mymin=INT_MAX;

    if (*p != BOXSIZE2D) {
        error(where,"BOXSIZE differs...");
        return;
    }

    sprintf(scrit.station,"%s",*station);
    scrit.t_start = (time_t) *start;
    scrit.t_end = (time_t) *end;
    printf(">> %d %d\n",(int) scrit.t_start,(int) scrit.t_end);

    if ((noobs=read123(*infile, &scrit, &asdata, &nsdata, &sobs)) <= 0) {
        error(where," File trouble...");
    }
    if (noobs != *n) {
        error(where,"Allocated space differs...");
        printf(" %d %d\n", noobs, *n);
        return;
    }

    /*
     * Retrieve data for analysis within R
     */
    j=0;
    for (k=0;k<noobs;k++) {
        ut.tm_year = ((sobs[k].year)-1900);
        ut.tm_mon = ((sobs[k].month)-1);
        ut.tm_mday = sobs[k].day;
        ut.tm_hour = sobs[k].hour;
        ut.tm_min = sobs[k].min;
        ut.tm_sec = 0;
        ut.tm_isdst = 0;
        vt = mktime(&ut);
        /*
           printf("%03d %d%02d%02d %02d:%02d - %d\n",
           k,
           sobs[k].year,sobs[k].month,sobs[k].day,
           sobs[k].hour,sobs[k].min,
           vt);
           */
        if (vt<mymin) {
            mymin=vt;
        }
        if (vt>mymax) {
            mymax=vt;
        }
        for (i=0;i<asdata[k].nopix;i++) {
            stid[j] = atoi(sobs[k].stID);
            mytime[j] = (int) vt;
            N[j] = (int) sobs[k].N;
            CL[j] = (int) sobs[k].Cl;
            CM[j] = (int) sobs[k].Cm;
            CH[j] = (int) sobs[k].Ch;
            E[j] = (int) sobs[k].E;
            sss[j] = (int) sobs[k].sss;
            k1[j] = (double) asdata[k].ch1[i];
            k2[j] = (double) asdata[k].ch2[i];
            k3a[j] = (double) asdata[k].ch3a[i];
            k3b[j] = (double) asdata[k].ch3b[i];
            k4[j] = (double) asdata[k].ch4[i];
            k5[j] = (double) asdata[k].ch5[i];
            soz[j] = (double) asdata[k].ang.soz;
            saz[j] = (double) asdata[k].ang.saz;
            raz[j] = (double) asdata[k].ang.raz;
            t0m[j] = (double) nsdata[k].t0m[0];
            t2m[j] = (double) nsdata[k].t2m[0];
            j++;
        }
    }
    /*
       printf(" Max utime: %d\n Min utime: %d\n",(int) mymax,(int) mymin);
       */

    return;
}

#endif
